lib/Dialect/FIRRTL/InnerSymbolTable.cpp

```diff
@@ -51,6 +52,19 @@ Operation *InnerSymbolTable::lookup(StringAttr name) const {
   return symbolTable.lookup(name);
 }
 
+/// Get InnerSymbol for an operation.
+StringAttr InnerSymbolTable::getInnerSymbol(Operation *op) {
+  return op->getAttrOfType<StringAttr>(
+      InnerSymbolTable::getInnerSymbolAttrName());
+}
```

Comment:
What happens when the `inner_sym` does not exist ? Do we need a `getOrAddInnerSymbol` ?

Reply:
Yes! We could just bail out to a lambda for that for the caller to provide to get this going now, but this seems something that a populated InnerSymbolTable is in a position to handle reasonably well for us and so perhaps we can do better.

To get there:

1) Extend this beyond Operations, to include ports (at least).  Few ways to hook/implement this (ideally in a way that introducing new targetable things isn't too invasive).  This gets more complicated if InnerSymbols point to entities within an Operation, so some sufficiently flexible design is needed here.  (Presumably the InnerSymbolTable should be able to `lookup` all InnerSymbol's)
2) Determine a method for generating suitable names when the hint/etc isn't available, similar to what we use `ModuleNamespace::newName` for presently.  This doesn't have to be complicated, but I'm not sure if we want to preserve what happens today re:per-name tracking of integer suffix.

And looking forward, sketch of what I'm thinking:

3) Drop ModuleNamespace for InnerSymbolTable, at least in FIRRTL.
4) Add methods for manipulating/managing an InnerSymbolTable so it can be preserved where possible (and make it suitable for use as an analysis).
5) Maybe replace the pattern of ModuleNamespace lazy creation with InnerSymbolTableCollection ?

Reply:
This sounds like a nice plan, should cleanup and refactor a lot of our inner sym related implementation.